### DISCUSSION TITLE
Websocket documentation error ?

### DISCUSSION THREAD

**Discussion by dev31337** [cesanta/mongoose#2408] [Ask questions, get answers] [answered]
Hi,

The websocket documentation [here](https://mongoose.ws/documentation/#struct-mg_ws_message) says that the flags element corresponds to the first byte as described in RFC 6455 section 5.2.

But when we look at the RFC the opcode is bits 4,5,6,7 while with the flags the opcode is bits 0,1,2,3.

The example makes it pretty clear by doing "msgtype = wm->flags & 0x0F;" to get the opcode.

This brings another question, if the flags element does not contain the FIN bit, that means we can not detect fragmented websocket messages,

Does that mean that Mongoose automatically buffers and de-fragments a fragmented websocket message and thus only issue one MG_EV_WS_MSG event when the final fragment is received ?

Regards.

**Comment by scaprile** [comment, marked as answer]
https://datatracker.ietf.org/doc/html/rfc6455#section-5.2 :
> When encoded on the wire, the most significant bit is the leftmost in the ABNF

`msg->flags` is not the actual byte in the WS frame, but a copy of it. The order of the bytes in the frame is as seen left to right, and the order of the bits is MSb to LSb left to right. The docs explain `msg->flags` and that is nevertheless coincident with that.

We don't strip anything, you just don't need to use that, because:
https://github.com/cesanta/mongoose/blob/cb1ae312a17a777a3e5c3b0db98be9a1f7232b1e/src/ws.c#L72
https://github.com/cesanta/mongoose/blob/cb1ae312a17a777a3e5c3b0db98be9a1f7232b1e/src/ws.c#L179
https://github.com/cesanta/mongoose/blob/cb1ae312a17a777a3e5c3b0db98be9a1f7232b1e/src/ws.c#L211-L230
Mongoose buffers the fragments

**Reply by dev31337** [reply]
Thank you for this very detailed answer ! I understand now...

This is really nice that Mongoose handles fragmented messages by itself, may I suggest that you tell it in the documentation because it is a great feature.
